### ISSUE TITLE
Question about dataset config and reproducing results

### ISSUE THREAD

**Issue by ChenBarryHu** [yuxumin/PoinTr#51]
Hi Xumin,

Congrats on the great work and thank you for sharing your work with the community!

I have two questions, could you help me? 

1. under the PoinTr.yaml, why val dataset is using "test" as a subset? (see the red circle in the picture below)
<img width="858" alt="Screen Shot 2022-05-06 at 9 34 35 PM" src="https://user-images.githubusercontent.com/17330047/167206252-a02950c8-ebf9-4da2-b52b-390c23fa30ef.png">

2. I am trying to reproduce the paper result of the PoinTr model on PCN dataset. I am using a single 3080Ti GPU, and I set the batch size to 26. In addition, I changed the "test" circled in the picture above into "val". Other than that, everything is following the configuration of the repo. 

Here is my training curves for about 200 epoches):
![Screenshot from 2022-05-06 21-46-04](https://user-images.githubusercontent.com/17330047/167207590-a4730f12-db4f-4bff-b1c3-fdfb3b8a6aad.png)

Somehow, the validation curves seem to be not stable, is this normal?

Looking forward to your reply, and have a great weekend!

Barry,
Technical University of Munich

**Comment by yuxumin** [comment]
> Somehow, the validation curves seem to be not stable, is this normal?

I think this problem is caused by the small batch size.

> under the PoinTr.yaml, why val dataset is using "test" as a subset? (see the red circle in the picture below)

Yes, using validation set to pick out the best model during the training is right. I believe you can easily reproduce the results in our paper, not matter `test` or `val` set is used in validation phase

**Comment by ChenBarryHu** [comment]
Hi Xumin,

Thanks for the fast reply!  However, the validation loss is calculated once at the end of each epoch, and the loss is aggregated using the whole validation set. Thus, in my opinion, a smaller batch size should not cause the instability of validation curves.

When you trained the model on PCN dataset, was the validation curve for "Loss/Epoch/Dense" going down steadily? Or it was also fluatuating?

Regards,
Barry

**Comment by yuxumin** [comment]
A smaller batch size will influence the optimization process during the training. In testing, the batch size is set to 1 in fact https://github.com/yuxumin/PoinTr/blob/70a2918e01b4cb335c6e68a56783941d449218e5/tools/builder.py#L25

**Comment by ChenBarryHu** [comment]
True, you are right. 

Btw, if setting "step_per_update" to "2", does it mean that the gradient descent would happen once every two batch? Like the losses calculated from two consecutive batches would be aggregated together then apply an optimization step?

-Barry

**Comment by yuxumin** [comment]
Yes, that is. The gradient will be accumulated every two batches. There is some little difference with bigger bs, however. But i think it will make the model optimization more stable. Take a try, and feel free to update the results here :)

**Comment by ChenBarryHu** [comment]
Thank you so much, Xumin! I will try it and share the results :)

**Comment by ChenBarryHu** [comment]
Sorry for bugging you with another message :) 
Theoretically, if setting "step_per_update" to "2", does it make sense to halve the initial learning rate from 0.0005 to 0.00025? Thanks!

**Comment by yuxumin** [comment]
According to the experience in 2D research. LR and BS should be proportional. In the default config, we set lr 5e-4 with bs 48 and update freq = 1. In your situation, i think the LR can be calculated out with this experience.

Reference:  https://github.com/facebookresearch/mae/blob/main/main_finetune.py#L267

**Comment by ChenBarryHu** [comment]
Thanks a lot!

**Comment by ChenBarryHu** [comment]
Tried to train with "step_per_update" to "2", the loss blows up at epoch 77:
![Screenshot from 2022-05-08 10-04-53](https://user-images.githubusercontent.com/17330047/167287374-00c49723-a349-4294-bdab-e31310679a10.png)

**Comment by yuxumin** [comment]
emmm, try to drop some backwards when `grad_norm` is larger/stranger compared with several previous steps?

**Comment by ChenBarryHu** [comment]
Yes, I am gonna apply `clip_gradient` and see how the training looks like, will keep you updated. Thx!

**Comment by ChenBarryHu** [comment]
Hi Xumin,

with `clip_gradient`, the curves behave normally, and I am able to reproduce good results:
![Screenshot from 2022-05-09 13-29-14](https://user-images.githubusercontent.com/17330047/167401049-583df6a8-47d7-4c67-a56d-f500a5df4b66.png)

Thanks for all your help!
Barry

**Comment by lin-lin-1003** [comment]
> Hi Xumin,
> 
> with `clip_gradient`, the curves behave normally, and I am able to reproduce good results: ![Screenshot from 2022-05-09 13-29-14](https://user-images.githubusercontent.com/17330047/167401049-583df6a8-47d7-4c67-a56d-f500a5df4b66.png)
> 
> Thanks for all your help! Barry

Could you please share how to adjust the code? Thank you very much
